Made continuous RL environments' API consistent

favre49 opened the pull request:
I also added an isTerminal() method for Pendulum environment

zoq commented on src/mlpack/methods/reinforcement_learning/environment/pendulum.hpp:
```diff
@@ -93,13 +93,17 @@ class Pendulum
            const double maxTorque = 2.0,
-           const double dt = 0.05) :
+           const double dt = 0.05,
+           const double angleThreshold = M_PI/6) :
```
Is that some default value you selected? Also there is a missing space between the ``/`` operator.

favre49 replied:
Yes, is this too lenient? perhaps `M_PI/12` is better?

zoq replied:
I was mainly just wondering where this value comes from, but I agree ``M_PI/12`` might be better.

favre49 replied:
Also, should I update HISTORY.MD? We didn't do it for PR #1931 so I'm not sure when we do this?

zoq replied:
Good point, I forgot about the HISTORY document, would be great to update the file as well.

manish7294 replied:
I wonder why do we need to impose constraint on pendulum using some angleThreshold, Isn't it better to just have terminals at `-PI `& `PI`, otherwise we won't ever be exploring with those states? Moreover this threshold still seems to be a bit large to me. Does it make sense?

zoq replied:
The idea here is that everything below the threshold would result in a terminal state. So we should choose something small, the issue I see here is that we don't terminate if we are above the threshold. So I guess it would make sense to introduce a maximal number of steps? I guess in this case we could also use the maximal number of steps as the terminal check?

favre49 replied:
> the issue I see here is that we don't terminate if we are above the threshold

I'm not sure I understand, have I made a mistake in my terminal condition?

>  So I guess it would make sense to introduce a maximal number of steps?

I agree. I think if we do that for the Pendulum environment, we should do this for all environments.

manish7294 replied:
> I'm not sure I understand, have I made a mistake in my terminal condition?
No, there is notthing wrong there. As per my understanding Marcus wants to highlight the point that, as of now we don't terminate even when we reaches the terminal state, we just now have a function to lookup whether we arein terminal state or not. So, It could be a good to introduce maximum number of steps to act as a factor for deciding termination point.

> I agree. I think if we do that for the Pendulum environment, we should do this for all environments.
 Sounds good. If we all agree on this, than maybe the change can just be done within this PR.

zoq replied:
> > the issue I see here is that we don't terminate if we are above the threshold
> 
> I'm not sure I understand, have I made a mistake in my terminal condition?

No issue with the. terminal condition, just an issue at which point we stop, but that applies for all tasks.

>I agree. I think if we do that for the Pendulum environment, we should do this for all environments.
> > Sounds good. If we all agree on this, than maybe the change can just be done within this PR.

Fine with me, but I think the default should either be no limit or a pretty high number. What do you think?

favre49 replied:
I think we could make the default value 0, in which case we won't make a check for termination.

favre49 replied:
Do you think we should maybe add a `Log::Info` stating the reason for termination of the episode, since now there are two possibilities?

zoq replied:
Good idea.

favre49 commented:
I fixed the multiple pole cart environments, they were meant to have one column more in `data`. I've changed the for loops appropriately as well.

I also made the changes we talked about in the Pendulum environment. I also noticed the environment does not include a `doneReward`, so I included that as well. If how I have implemented this in the Pendulum environment is fine, I will extend it to the other environments as well.

Also, since the change in the `Action` struct would qualify as non-backwards compatible, should I make a change in HISTORY.md to include these in 4.0.0?

zoq commented on src/mlpack/methods/reinforcement_learning/environment/pendulum.hpp:
```diff
@@ -93,13 +93,24 @@ class Pendulum
+           const double dt = 0.05,
+           const double angleThreshold = M_PI / 12,
+           const double doneReward = 0.0,
+           const size_t maxTimeSteps = 0) :
```
Not sure about the name, do you think ``maxSteps`` or ``maxIterations`` could work as well?

zoq commented:
About the backwards compatible, I think that is reasonable, don't see an easy workaround to support both.

manish7294 commented on src/mlpack/methods/reinforcement_learning/environment/pendulum.hpp:
```diff
@@ -140,6 +154,15 @@ class Pendulum
+    bool done = IsTerminal(nextState);
+
+    // Do not reward the agent if time ran out.
+    if (done && maxTimeSteps != 0 && timeStepsPerformed >= maxTimeSteps)
```
Should we terminate the moment we reached maximum iterations allowed? Let me know your thoughts on this?

manish7294 replied:
Oh, I see you are doing the same. I thought done is being taken from angle terminal condition. Ignore the above comment.

zoq commented on src/mlpack/tests/rl_components_test.cpp:
```diff
@@ -37,7 +37,7 @@ BOOST_AUTO_TEST_SUITE(RLComponentsTest)
 BOOST_AUTO_TEST_CASE(SimplePendulumTest)
 {
-  const Pendulum task = Pendulum();
+  Pendulum task = Pendulum();
```
Might be a good idea to test terminates after the maximal steps as well, we could use a super low value, to keep the runtime low.

favre49 commented:
I extended what I did with `Pendulum` to other environments, and added tests as well. Besides that, I noticed some environments don't have `doneReward`, so I added that. Finally, I removed the `typedef` from `Acrobot` seeing as this PR will remove backwards compatibility anyway.

zoq commented:
Great, thanks for looking into it.

zoq commented on src/mlpack/methods/reinforcement_learning/environment/acrobot.hpp:
```diff
@@ -192,20 +204,33 @@ class Acrobot
    * This function checks if the acrobot has reached the terminal state.
    *
    * @param state The current State.
+   * @return true if state is a terminal state, otherwise false.   * 
```
We can remove the ``*`` at the end.

zoq commented on src/mlpack/methods/reinforcement_learning/environment/pendulum.hpp:
```diff
@@ -93,13 +93,25 @@ class Pendulum
+   * @param angleThreshold The region about the upright position where the
+   *    state is considered terminal.
+   * @param doneReward The reward recieved by the agent on success.
+   * @param maxTimeSteps The number of time steps after which the episode
```
Can you change every occurrence of ``maxTimeSteps`` with ``maxSteps``?

zoq commented on src/mlpack/methods/reinforcement_learning/environment/pendulum.hpp:
```diff
@@ -140,6 +155,15 @@ class Pendulum
+    bool done = IsTerminal(nextState);
+
+    // Do not reward the agent if time ran out.
+    if (done && maxSteps != 0 && timeStepsPerformed >= maxSteps)
```
Since we are going with ``maxSteps`` I think it makes sense to call this ``stepsPerformed``?

zoq commented on src/mlpack/methods/reinforcement_learning/environment/pendulum.hpp:
```diff
@@ -183,6 +208,32 @@ class Pendulum
+  {
+    if (maxSteps != 0 && timeStepsPerformed >= maxSteps)
+    {
+      Log::Info << "Episode terminated due to the maximum number of time steps"
```
Should we go with ``number of steps``?

zoq commented on src/mlpack/methods/reinforcement_learning/environment/pendulum.hpp:
```diff
@@ -183,6 +208,32 @@ class Pendulum
+  }
+
+  //! Get the number of time steps performed
+  size_t TimeStepsPerformed() const { return timeStepsPerformed; }
```
See comment above.

zoq commented on src/mlpack/tests/rl_components_test.cpp:
```diff
@@ -37,7 +37,7 @@ BOOST_AUTO_TEST_SUITE(RLComponentsTest)
 BOOST_AUTO_TEST_CASE(SimplePendulumTest)
 {
-  const Pendulum task = Pendulum();
+  Pendulum task = Pendulum(8, 2, 0.05, M_PI/12, 0, 5);
```
Picky style comment, missing space between ``/``.

zoq commented on src/mlpack/tests/rl_components_test.cpp:
```diff
@@ -57,15 +65,24 @@ BOOST_AUTO_TEST_CASE(SimplePendulumTest)
 BOOST_AUTO_TEST_CASE(SimpleContinuousMountainCarTest)
 {
-  const ContinuousMountainCar task = ContinuousMountainCar();
+  ContinuousMountainCar task = ContinuousMountainCar(-1.2, 0.6, 0.45, -0.07,
```
Might be a good idea to introduce a method to set the maximal number of steps? That way we could use the default constructor.

manish7294 commented on src/mlpack/methods/reinforcement_learning/environment/cart_pole.hpp:
```diff
@@ -104,7 +104,9 @@ class CartPole
    * @param thetaThresholdRadians The maximum angle.
    * @param xThreshold The maximum position.
-   * @param doneReward Reward recieved on termination.
+   * @param doneReward Reward recieved by agent on success.
```
picky comment!, I think we should keep the description of doneReward as "Reward recieved by agent on termination"

favre49 replied:
I made this change since now the termination of an episode can happen because the maximum number of steps have been taken (which could be a success in cases like cart pole but a failure because in other cases) or any other termination case we have included. I changed the description to make it more clear, since it may not be evident to the user.

manish7294 replied:
Still I think termination word deals with both the cases unbiasedly. Though, no worries, you can keep it that way. :)

manish7294 commented:
Great work, glad to merge this in.
